frontend/src/MapContainer.tsx

```diff
@@ -0,0 +1,117 @@
+    props.eventBus.register(EventBusEventType.USER_CHANGE, (e, u) => this.refreshState(u));
+    L.Icon.Default.imagePath="icons/map/";
+  }
+
+  /** Called when the user clicks on the map to update the pin. */
+  private onMapClick(e: any) {
+    this.setState({
+      lat: e.latlng.lat,
+      lng: e.latlng.lng,
+      changed: true,
+      refreshing: false
+    });
+  }
+
+  /** Called when the user zooms to avoid resetting the zoom after saving. */
+  private onMapZoom(e: any) {
+    this.setState({ zoom: this.map ? this.map.leafletElement.getZoom() : this.state.zoom });
+  }
+
+  /**Refreshes the entire map based on the new login status information. */
+  private refreshState(loginStatus: LoginStatus) {
+    let loc = UserConfigurationDecoder.decodeLocationFromLoginStatus(loginStatus);
+    let lat = loginStatus.config.find(x => x.descriptor === "user.latitude")?.value;
+    let lng = loginStatus.config.find(x => x.descriptor === "user.longitude")?.value;
+    this.setState({ location: loc, zoom: this.state.zoom, lat: lat ? Number(lat) : 50, lng: lng ? Number(lng) : 10, changed: false, refreshing: false });
+  }
+
+  /** Persist the selected location to the backend. Updates the map and e.g. the trending twitter topics. */
+  private saveSelectedLocation(e: any) {
+    this.setState({ refreshing: true });
+    fetch(`${DeploymentManager.getUrl()}api/user/location`, {
+      method: 'PUT',
+      headers: new Headers({
+        'Content-Type': 'application/json'
+      }),
+      credentials: 'include',
+      body: JSON.stringify({ latitude: this.state.lat, longitude: this.state.lng })
+    })
+      .then(response => {
+        if (!response.ok) {
+          ToastManager.showErrorToast(response);
+        } else {
+          response.json().then(data => {
+            ToastManager.showSuccessToast("Updated location to " + data.locationName + ".");
+            setTimeout(() => this.props.loginManager.logIn(), 2500);
+          });
+        }
+      });
+  }
+
+  render() {
+    return (
+      <div className="container double-container inline-block">
+        <div className="box-header">
+          Select your Location
+                </div>
+        <div className="box-content">
+          {this.state.refreshing ? <div>Refreshing your current location...</div> : <span>
+            <div className="displayName display-name-bold inline-block">Your currently selected location:&nbsp;</div>
+            <span>{this.state.location.name}, {this.state.location.placeType.name} in {this.state.location.country}</span>
+          </span>}
+          <div className="mapContainer">
+            <Map ref={ref => this.map = ref} onzoomlevelschange={e => this.onMapZoom(e)} onzoomend={e => this.onMapZoom(e)} onzoomstart={e => this.onMapZoom(e)} onclick={e => this.onMapClick(e)} className="mapContainer" center={this.state} zoom={this.state.zoom}>
```

Comment:
Maybe use the newer [`React.createRef()`](https://reactjs.org/docs/react-api.html#reactcreateref) instead.

Reply:
Interesting. Not sure if the Map-API supports this though...

Reply:
Not sure, if Map-API decides that. From what I read, it's more about being a class or function component.